# Task: Linear vs Polynomial Regression

## Objective
Build and compare Linear Regression and Polynomial Regression models then select the best performing model and interpret its learned equation.


## Dataset
Use the **Boston Housing dataset**.
    
    from sklearn.datasets import load_boston

In [3]:
import pandas as pd
from sklearn.datasets import fetch_openml

# Load Boston Housing dataset
boston = fetch_openml(name="boston", version=1, as_frame=True)

# Features and target
X = boston.data
y = boston.target.astype(float)

# Quick look
print(X.head())
print(y.head())
print(X.shape, y.shape)

C:\Users\Zeyad\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:61: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (
C:\Users\Zeyad\anaconda3\lib\site-packages\sklearn\datasets\_openml.py:932: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


      CRIM    ZN  INDUS CHAS    NOX     RM   AGE     DIS RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31    0  0.538  6.575  65.2  4.0900   1  296.0     15.3   
1  0.02731   0.0   7.07    0  0.469  6.421  78.9  4.9671   2  242.0     17.8   
2  0.02729   0.0   7.07    0  0.469  7.185  61.1  4.9671   2  242.0     17.8   
3  0.03237   0.0   2.18    0  0.458  6.998  45.8  6.0622   3  222.0     18.7   
4  0.06905   0.0   2.18    0  0.458  7.147  54.2  6.0622   3  222.0     18.7   

        B  LSTAT  
0  396.90   4.98  
1  396.90   9.14  
2  392.83   4.03  
3  394.63   2.94  
4  396.90   5.33  
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: MEDV, dtype: float64
(506, 13) (506,)


### 1. Data Preparation

In [4]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

# Check data info
print("Missing values in X:")
print(X.isnull().sum())
print("\nMissing values in y:")
print(y.isnull().sum())


Missing values in X:
CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

Missing values in y:
0


In [5]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [6]:

# Scale features for linear regression
scaler_linear = StandardScaler()
X_train_scaled = scaler_linear.fit_transform(X_train)
X_test_scaled = scaler_linear.transform(X_test)


In [7]:
# Create polynomial features
poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(X_train_scaled)
X_test_poly = poly.transform(X_test_scaled)

In [8]:
# Check shapes
print("Original X_train shape:", X_train.shape)
print("Scaled X_train shape:", X_train_scaled.shape)
print("Polynomial X_train shape:", X_train_poly.shape)

Original X_train shape: (404, 13)
Scaled X_train shape: (404, 13)
Polynomial X_train shape: (404, 104)


### 2. Linear Regression Model
- Train a Linear Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE

In [9]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error

# Train Linear Regression model
linear_model = LinearRegression()
linear_model.fit(X_train_scaled, y_train)



LinearRegression()

In [10]:

# Make predictions
y_pred_linear = linear_model.predict(X_test_scaled)

In [11]:

# Evaluation metrics
mae = mean_absolute_error(y_test, y_pred_linear)
mse = mean_squared_error(y_test, y_pred_linear)
rmse = np.sqrt(mse)

# Print results
print("Linear Regression Results:")
print(f"MAE  : {mae:.4f}")
print(f"MSE  : {mse:.4f}")
print(f"RMSE : {rmse:.4f}")

Linear Regression Results:
MAE  : 3.1891
MSE  : 24.2911
RMSE : 4.9286


### 3. Polynomial Regression Model
- Train a Polynomial Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE


In [12]:
# Train Polynomial Regression model
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)



LinearRegression()

In [13]:

# Make predictions
y_pred_poly = poly_model.predict(X_test_poly)

In [14]:

# Evaluation metrics
mae_poly = mean_absolute_error(y_test, y_pred_poly)
mse_poly = mean_squared_error(y_test, y_pred_poly)
rmse_poly = np.sqrt(mse_poly)

# Print results
print("Polynomial Regression Results:")
print(f"MAE  : {mae_poly:.4f}")
print(f"MSE  : {mse_poly:.4f}")
print(f"RMSE : {rmse_poly:.4f}")

Polynomial Regression Results:
MAE  : 2.5722
MSE  : 14.2490
RMSE : 3.7748


### 4. Model Comparison
- Compare Linear vs Polynomial Regression results
- Identify which model performs better on the test data and justify your decision


In [15]:
# Create comparison table
comparison = pd.DataFrame({
    "Model": ["Linear Regression", "Polynomial Regression"],
    "MAE": [mae, mae_poly],
    "MSE": [mse, mse_poly],
    "RMSE": [rmse, rmse_poly]
})

print(comparison)

                   Model       MAE        MSE      RMSE
0      Linear Regression  3.189092  24.291119  4.928602
1  Polynomial Regression  2.572160  14.249047  3.774791


In [16]:
if rmse_poly < rmse:
    print("Polynomial Regression performs better on the test data.")
    print("Justification: it has lower MAE, MSE, and RMSE, which means its predictions are closer to the true house prices.")
else:
    print("Linear Regression performs better on the test data.")
    print("Justification: it has lower MAE, MSE, and RMSE, so it generalizes better and the polynomial model likely added unnecessary complexity.")

Polynomial Regression performs better on the test data.
Justification: it has lower MAE, MSE, and RMSE, which means its predictions are closer to the true house prices.


The Polynomial Regression model performs better than the Linear Regression model on the test data, as it achieves lower error across all evaluation metrics. Its MAE decreases from 3.19 to 2.57, its MSE drops from 24.29 to 14.25, and its RMSE improves from 4.93 to 3.77, showing that its predictions are consistently closer to the actual house prices. This suggests that the relationship between the input features and housing prices is not purely linear, and that the polynomial model is better able to capture the underlying nonlinear patterns in the dataset.
